Use the new CompositeResourceDefinition type

muvaf opened the pull request:
Fixes #13

muvaf committed:
go.mod: update crossplane and crossplane-runtime to master

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
update codebase to use CompositeResourceDefinition type and remove usage of InfrastructurePublication

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

negz commented on cmd/agent/local/local.go:
```diff
@@ -28,7 +28,7 @@ import (
 	"github.com/crossplane/crossplane/apis/apiextensions"
 
-	"github.com/crossplane/agent/pkg/controllers/publication"
+	"github.com/crossplane/agent/pkg/controllers/xrd"
```
The controller for the corresponding type upstream is named `definition` - I think we should be consistent here. I'm happy if we rename both controller packages to `xrd`, though.

muvaf replied:
In this repo, there is a lot of operations going on with both CRDs and XRDs so I'm hesitant to call any of them `definition`. In fact, we have `pkg/controllers/crd` and `pkg/controllers/xrd` packages. So, I'd opt for `xrd` at least for this repo.

negz replied:
Is that not also true of the c/c apiextensions controllers?

muvaf replied:
> Is that not also true of the c/c apiextensions controllers?

Do you mean that the `apiextensions` controllers in the upstream could also have that separation? In this case, the `pkg/controllers/crd` actually watches the type `CustomResourceDefinition`. I don't think we have a controller in c/c that watches for `CustomResourceDefinition`s.

negz commented on pkg/controllers/apiextensions/setup.go:
```diff
@@ -30,27 +30,26 @@ import (
 
-	infraPubCRDName    = "infrastructurepublications.apiextensions.crossplane.io"
-	infraDefCRDName    = "infrastructuredefinitions.apiextensions.crossplane.io"
+	xrdCRDName         = "compositeresourcedefinition.apiextensions.crossplane.io"
```
```suggestion
	xrdCRDName         = "compositeresourcedefinitions.apiextensions.crossplane.io"
```

I think this would be plural, right?

negz commented on pkg/controllers/xrd/reconciler.go:
```diff
@@ -41,9 +42,9 @@ import (
 	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
 	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1/ccrd"
-	crequirement "github.com/crossplane/crossplane/pkg/controller/apiextensions/requirement"
+	cclaim "github.com/crossplane/crossplane/pkg/controller/apiextensions/claim"
```
Nit: These `cclaim` and `rresource` names threw me off when reviewing the original agent PR, because they look a lot like typos.

muvaf replied:
Changed with `coreclaim` and `runtimeresource`

muvaf committed:
remove condition check on predicate of xrd with claim filtering

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

negz commented on pkg/resource/predicates.go:
```diff
@@ -75,3 +78,47 @@ func (f *NameFilter) Generic(e event.GenericEvent) bool {
+}
+
+// XRDWithClaim only allows CompositeResourceDefinitions which offer a resource claim.
+type XRDWithClaim struct{}
```
Should we use https://godoc.org/sigs.k8s.io/controller-runtime/pkg/predicate#NewPredicateFuncs here?

negz reviewed (commented):
LGTM so far. I won't hit approve just yet given that the PR is still in draft.

muvaf committed:
rename CRDRenderers to CRDFetchers

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
add explanatory comments to local XRD reconciler

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
claim reconciler: add comments and separate Propagator interface into Configurator and Propagator

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
keep local and remote words only to clients

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
expose metrics of both agents in the Deployment

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
replace rresource with runtimeresource and make condition reasons to be short parseable strings

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
use predicate.Funcs from controller-runtime instead of implementing full Predicate structs

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf committed:
replace cclaim with coreclaim

Signed-off-by: Muvaffak Onus <onus.muvaffak@gmail.com>

muvaf commented:
I have tested this with Wordpress and `master` Crossplane and it's all good. @negz @kasey Thanks for your reviews, addressed your comments in the old PR as well.

muvaf commented:
@negz @kasey  could you take a quick look? There isn't really big changes besides some simplifications that address your concerns.

negz commented on cmd/agent/remote/remote.go:
```diff
@@ -48,7 +47,7 @@ func (a *Agent) Run(log logging.Logger, period time.Duration) error {
 	}
 
-	mgr, err := ctrl.NewManager(a.ClusterConfig, ctrl.Options{SyncPeriod: &period, MetricsBindAddress: "0"})
+	mgr, err := ctrl.NewManager(a.ClusterConfig, ctrl.Options{SyncPeriod: &period, MetricsBindAddress: "127.0.0.1:8081"})
```
Do we really want to listen for metrics only on localhost?

muvaf replied:
Just changed it to `0.0.0.0`. That should enable it to be accessible from outside of the `Pod`.

negz commented on pkg/controllers/xrd/reconciler.go:
```diff
@@ -52,34 +53,36 @@ const (
 	localPrefix        = "local cluster: "
 	remotePrefix       = "remote cluster: "
-	errUpdateStatus    = "cannot update status of infrastructure publication"
+	errUpdateStatus    = "cannot update status of xrd"
```
Nit: These instances of `xrd` should probably be `composite resource definition` (we write out `custom resource definition`, or `XRD`.
